File: 백준/Gold/1644. 소수의 연속합/소수의 연속합.py
import sys
import math
input = sys.stdin.readline

def isPrimeNum(x):
    for i in range(2, int(math.sqrt(x)) + 1):
        if x % i == 0:
            return False
    return True

n = int(input())

if n == 1:
    print(0)
    exit()

primes = []

for i in range(2, n + 1):
    if isPrimeNum(i):
        primes.append(i)

start, end = 0, 0
sum = primes[0]
cnt = 0

while start <= end and end < len(primes):
    if sum < n:
        end += 1
        if end < len(primes):
            sum += primes[end]
    elif sum == n:
        cnt += 1
        end += 1
        if end < len(primes):
            sum += primes[end]
    elif sum > n:
        sum -= primes[start]
        start += 1

        if start > end and start < len(primes):
            end = start
            sum = primes[start]

print(cnt)

File: 백준/Gold/1644. 소수의 연속합/소수의 연속합.cc
#include <iostream>
#include <math.h>
#include <vector>
#define MAX 2000

using namespace std;

bool isPrimeNum(int x) {
    int x_sqrt = (int)sqrt(x);
    for (int i = 2; i <= x_sqrt; i++) {
        if (x % i == 0)
            return false;
    }
    return true;
}

int main() {
    int n;
    cin >> n;

    if (n == 1) {
        cout << "0\n";
    } else {
        vector<int> primes;
        primes.reserve(n);
        int primes_length = 0;

        for (int i = 2; i < n + 1; i++) {
            if (isPrimeNum(i))
                primes.push_back(i);
        }

        int start = 0;
        int end = 0;
        int sum = 2;
        int cnt = 0;

        while (start <= end && end < primes.size()) {
            if (sum < n) {
                sum += primes[++end];
            } else if (sum == n) {
                cnt += 1;
                sum += primes[++end];
            } else if (sum > n) {
                sum -= primes[start++];

                if (start > end && start < primes.size()) {
                    end = start;
                    sum = primes[start];
                }
            }
        }

        cout << cnt << "\n";
    }

    return 0;
}
